lib/astarte_data_access/xandra_utils.ex

```diff
@@ -23,67 +23,120 @@ defmodule Astarte.DataAccess.XandraUtils do
+
+      {:error, %Xandra.ConnectionError{} = err} ->
+        _ =
+          Logger.warning("Database connection error: #{inspect(err)}.",
+            tag: "database_connection_error"
+          )
+
+        {:error, :database_connection_error}
+
+      other ->
+        other
     end
   end
 
-  defp verify_realm(realm_name) do
-    case Realm.valid_name?(realm_name) do
-      true -> :ok
-      false -> {:error, :invalid_realm_name}
+  def validated_run(realm_name, fun, opts \\ []) do
+    with :ok <- validate_realm_name(realm_name),
+         {:ok, result} <- run(realm_name, fun, opts) do
+      {:ok, result}
     end
   end
 
-  @spec retrieve_page(
-          atom | pid | {atom, any} | {:via, atom, any} | DBConnection.t(),
-          binary | Xandra.Prepared.t(),
-          list | map,
-          keyword
-        ) ::
-          {:error, :database_error}
-          | {:error, :database_connection_error}
-          | {:ok, Xandra.Page.t()}
-  def retrieve_page(conn, statement, params, opts \\ []) do
-    with {:ok, prepared} <- prepare_query(conn, statement) do
-      opts = opts ++ default_query_opts()
+  def execute_query(conn, statement), do: execute_query(conn, statement, %{}, [])
+
+  def execute_query(conn, statement, opts) when is_list(opts),
+    do: execute_query(conn, statement, %{}, opts)
+
+  def execute_query(conn, statement, params) when is_map(params),
+    do: execute_query(conn, statement, params, [])
+
+  def execute_query(conn, query, params, opts) when is_map(params) and is_list(opts) do
+    opts = Keyword.merge(page_default_opts(), opts)
 
+    with {:ok, prepared} <- prepare_query(conn, query, opts) do
       case Xandra.execute(conn, prepared, params, opts) do
+        {:ok, result} ->
+          {:ok, result}
+
         {:error, %Xandra.Error{} = reason} ->
-          _ = Logger.warn("Database error while retrieving data: #{inspect(reason)}")
+          _ = Logger.warning("Database error while executing: #{inspect(reason)}")
           {:error, :database_error}
 
         {:error, %Xandra.ConnectionError{} = reason} ->
-          _ = Logger.warn("Database connection error while retrieving data: #{inspect(reason)}")
+          _ = Logger.warning("Database connection error while executing: #{inspect(reason)}")
           {:error, :database_connection_error}
-
-        {:ok, %Xandra.Page{} = page} ->
-          {:ok, page}
       end
     end
   end
 
-  defp prepare_query(conn, statement, opts \\ []) do
+  def retrieve_page(conn, query), do: retrieve_page(conn, query, %{}, [])
+
+  def retrieve_page(conn, query, opts) when is_list(opts),
+    do: retrieve_page(conn, query, %{}, opts)
+
+  def retrieve_page(conn, query, params) when is_map(params),
+    do: retrieve_page(conn, query, params, [])
+
+  def retrieve_page(conn, query, params, opts) when is_map(params) and is_list(opts) do
+    with {:ok, %Xandra.Page{} = page} <- execute_query(conn, query, params, opts) do
+      {:ok, page}
+    end
+  end
```

Comment:
Let's keep the typespecs and use default arguments. This would remove the need for most of the previous function heads:
```suggestion
  @spec retrieve_page(
          atom | pid | {atom, any} | {:via, atom, any} | DBConnection.t(),
          binary | Xandra.Prepared.t(),
          list | map,
          keyword
        ) ::
          {:error, :database_error}
          | {:error, :database_connection_error}
          | {:ok, Xandra.Page.t()}
  def retrieve_page(conn, query, params \\ %{}, opts \\ [])  do
```